File: VeroFlutter/lib/services/via_cep_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ViaCepResponse {
  final String? logradouro;
  final String? bairro;
  final String? localidade;
  final String? uf;

  const ViaCepResponse({this.logradouro, this.bairro, this.localidade, this.uf});

  factory ViaCepResponse.fromJson(Map<String, dynamic> j) => ViaCepResponse(
        logradouro: j['logradouro'] as String?,
        bairro: j['bairro'] as String?,
        localidade: j['localidade'] as String?,
        uf: j['uf'] as String?,
      );
}

abstract class ViaCepService {
  static Future<ViaCepResponse> fetchAddress(String cep) async {
    final clean = cep.replaceAll(RegExp(r'\D'), '');
    final uri = Uri.parse('https://viacep.com.br/ws/$clean/json/');
    final response = await http.get(uri).timeout(const Duration(seconds: 8));

    if (response.statusCode == 200) {
      final data = json.decode(response.body) as Map<String, dynamic>;
      if (data['erro'] == true) throw Exception('CEP não encontrado.');
      return ViaCepResponse.fromJson(data);
    }
    throw Exception('Erro ao buscar CEP (${response.statusCode}).');
  }
}

File: VeroApp/Services/ViaCEPService.swift
import Foundation

struct ViaCEPResponse: Codable {
    let logradouro: String?
    let bairro: String?
    let localidade: String?
    let uf: String?
    let erro: Bool?
}

enum ViaCEPError: LocalizedError {
    case notFound
    case network(Error)

    var errorDescription: String? {
        switch self {
        case .notFound:      return "CEP não encontrado. Preencha os campos manualmente."
        case .network(let e): return "Erro de rede: \(e.localizedDescription)"
        }
    }
}

enum ViaCEPService {
    static func fetchAddress(cep: String) async throws -> ViaCEPResponse {
        let clean = cep.filter(\.isNumber)
        guard clean.count == 8,
              let url = URL(string: "https://viacep.com.br/ws/\(clean)/json/") else {
            throw ViaCEPError.notFound
        }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let response = try JSONDecoder().decode(ViaCEPResponse.self, from: data)
            if response.erro == true { throw ViaCEPError.notFound }
            return response
        } catch let e as ViaCEPError {
            throw e
        } catch {
            throw ViaCEPError.network(error)
        }
    }
}
